app/models/manageiq/providers/microsoft/infra_manager/scanning/job.rb

```diff
@@ -0,0 +1,173 @@
+class ManageIQ::Providers::Microsoft::InfraManager::Scanning::Job < VmScan
+  # Make updates to default state machine to take into account snapshots
+  def load_transitions
+    super.tap do |transitions|
+      transitions.merge!(
+        :start_snapshot     => {'before_scan'               => 'snapshot_create'},
+        :snapshot_complete  => {'snapshot_create'           => 'scanning',
+                                'snapshot_delete'           => 'synchronizing'},
+        :snapshot_delete    => {'scanning'                  => 'snapshot_delete'},
+        :data               => {'snapshot_create'           => 'scanning',
+                                'scanning'                  => 'scanning',
+                                'snapshot_delete'           => 'snapshot_delete',
+                                'synchronizing'             => 'synchronizing',
+                                'finished'                  => 'finished'}
+      )
+    end
+  end
+
+  def before_scan
+    signal(:start_snapshot)
+  end
+
+  def call_snapshot_create
+    _log.info("Enter")
+
+    begin
+      context[:snapshot_mor] = nil
+
+      options[:snapshot] = :skipped
+      options[:use_existing_snapshot] = false
+      return unless create_snapshot
+      signal(:snapshot_complete)
+    rescue => err
+      _log.log_backtrace(err)
+      signal(:abort, err.message, "error")
+      return
+    rescue Timeout::Error
+      msg = case options[:snapshot]
+            when :smartProxy, :skipped then "Request to log snapshot user event with EMS timed out."
+            else "Request to create snapshot timed out"
+            end
+      _log.error(msg)
+      signal(:abort, msg, "error")
+    end
+  end
+
+  def call_snapshot_delete
+    _log.info("Enter")
+
+    # TODO: remove snapshot here if Vm was running
+    vm = VmOrTemplate.find(target_id)
+    if context[:snapshot_mor]
+      mor = context[:snapshot_mor]
+      context[:snapshot_mor] = nil
+
+      if options[:snapshot] == :smartProxy
+        set_status("Snapshot delete was performed by the SmartProxy")
+      else
+        set_status("Deleting VM snapshot: reference: [#{mor}]")
+      end
+
+      if vm.ext_management_system
+        _log.info("Deleting snapshot: reference: [#{mor}]")
+        begin
+          delete_snapshot(mor, vm)
+        rescue => err
+          _log.error(err.to_s)
+          return
+        rescue Timeout::Error
+          msg = "Request to delete snapshot timed out"
+          _log.error(msg)
```

Comment:
Same issue here as I mentioned in the other one. Without a return, won't this look like it succeeded?

Reply:
@djberg96 I wonder how this worked prior to this refactoring.